src/CoreWCF.Primitives/src/CoreWCF/Channels/SecurityBindingElement.cs

```diff
@@ -298,52 +299,172 @@ private static BindingContext CreateIssuerBindingContextForNegotiation(BindingCo
+                return false;
+            if ((parameters.Signed.Count == 1) && (parameters.Endorsing.Count == 0))
             {
-                SetIssuerBindingContextIfRequired(supportingParameters.Endorsing[i], issuerBindingContext);
+                issuedTokenParameters = parameters.Signed[0] as IssuedSecurityTokenParameters;
+                if (issuedTokenParameters != null && issuedTokenParameters.KeyType != SecurityKeyType.BearerKey)
+                    return false;
             }
-            for (int i = 0; i < supportingParameters.SignedEndorsing.Count; ++i)
+            else if ((parameters.Endorsing.Count == 1) && (parameters.Signed.Count == 0))
             {
-                SetIssuerBindingContextIfRequired(supportingParameters.SignedEndorsing[i], issuerBindingContext);
+                issuedTokenParameters = parameters.Endorsing[0] as IssuedSecurityTokenParameters;
+                if (issuedTokenParameters != null && (issuedTokenParameters.KeyType != SecurityKeyType.SymmetricKey && issuedTokenParameters.KeyType != SecurityKeyType.AsymmetricKey))
+                    return false;
             }
-            for (int i = 0; i < supportingParameters.Signed.Count; ++i)
+            if (issuedTokenParameters == null)
+                return false;
+            if (issuedTokenParameters.RequireDerivedKeys)
+                return false;
+
+            return true;
+        }
+
+
+        // If any changes are made to this method, please make sure that they are
+        // reflected in the corresponding IsSslNegotiationBinding() method.
+        public static SymmetricSecurityBindingElement CreateSslNegotiationBindingElement(bool requireClientCertificate, bool requireCancellation)
+        {
+            SymmetricSecurityBindingElement result = new SymmetricSecurityBindingElement(
+                new SslSecurityTokenParameters(requireClientCertificate, requireCancellation));
+            return result;
+        }
+
+        public static SymmetricSecurityBindingElement CreateIssuedTokenBindingElement(IssuedSecurityTokenParameters issuedTokenParameters)
+        {
+            if (issuedTokenParameters == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(issuedTokenParameters));
+            if (issuedTokenParameters.KeyType != SecurityKeyType.SymmetricKey)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgument(SR.Format(SR.IssuedTokenAuthenticationModeRequiresSymmetricIssuedKey));
+            SymmetricSecurityBindingElement result = new SymmetricSecurityBindingElement(issuedTokenParameters);
+            return result;
+        }
+
+        // If any changes are made to this method, please make sure that they are
+        // reflected in the corresponding IsIssuedTokenForCertificateBinding() method.
```

Comment:
Remove comment about IsIssuedTokenForCertificate as that method isn't needed (client side usage only)

Reply:
This is called when the mode is Message(not Secure).. although the Message mode is not present, I thought of letting you know before removing